### PR TITLE
Implement polyfill for globalThis

### LINKED ISSUE #4350: iOS - Can't find variable: globalThis
### Description

Using `10.20.0-alpha.2` (but without actually using Hermes on iOS), reported the above crash for a user right on app startup.

### Stacktrace & log output

```shell
ReferenceError
Can't find variable: globalThis

    node_modules/realm-network-transport/dist/bundle.react-native.es.js:91:43 
    node_modules/metro-runtime/src/polyfills/require.js:340:6 loadModuleImplementation
    node_modules/realm/lib/extensions.js:60:45 exports
    node_modules/realm/lib/react-native.js:45:24 
    node_modules/metro-runtime/src/polyfills/require.js:340:6 loadModuleImplementation
    src/db/index.js:8 
    node_modules/metro-runtime/src/polyfills/require.js:340:6 loadModuleImplementation
    src/App.js:31 
    node_modules/metro-runtime/src/polyfills/require.js:340:6 loadModuleImplementation
    node_modules/metro-runtime/src/polyfills/require.js:200:45 guardedLoadModule
    index.js:86:29 
    node_modules/regenerator-runtime/runtime.js:63:44 tryCatch
    node_modules/regenerator-runtime/runtime.js:294:30 
    node_modules/regenerator-runtime/runtime.js:63:44 tryCatch
    node_modules/regenerator-runtime/runtime.js:155:28 invoke
    node_modules/regenerator-runtime/runtime.js:165:19 
    node_modules/promise/setimmediate/core.js:37:14 tryCallOne
    node_modules/promise/setimmediate/core.js:123:25 
    node_modules/react-native/Libraries/Core/Timers/JSTimers.js:248:17 
    node_modules/react-native/Libraries/Core/Timers/JSTimers.js:112:6 _callTimer
    node_modules/react-native/Libraries/Core/Timers/JSTimers.js:166:15 _callReactNativeMicrotasksPass
    node_modules/react-native/Libraries/Core/Timers/JSTimers.js:418:11 callReactNativeMicrotasks
    node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:393:11 value
    node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:135:11 
    node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:370:8 value
    node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:134:17 value
    [native code] value
    [native code] value
```
```


### Can you reproduce a bug?

Not yet

### Reproduction Steps

Regular offline (non sync) usage.

### Version

10.20.0-alpha.2

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

_No response_

### Platform OS and version(s)

iOS 12.1.1 (iPad Mini 3)

### PR DESCRIPTION
**Pull request by tomduncalf**
## What, How & Why?

This closes #4350. I decided to replace the existing inline implementations with a more robust polyfill taken from [core-js](https://github.com/zloirock/core-js/blob/master/packages/core-js/internals/global.js), and to put it in `realm.io/common` so we don't need to always repeat the same snippet

## ☑️ ToDos

* [x] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 🚦 Tests
* [ ] 🔀 Executed flexible sync tests locally if modifying flexible sync
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -181,6 +181,7 @@ const Person = {
 * Logging out too quickly can cause an error if the timeout behavior is set to `openLocalRealm` ([#4453](https://github.com/realm/realm-js/issues/4453), since v10.0.0)
 * Released `realm-network-transport` to adopt the changes published to fix `globalThis` undefined issue for older devices. ([#4350](https://github.com/realm/realm-js/issues/4350), since v10.0.0)
 * Fixed flexible sync crash when updating subscriptions after token expiry. ([#4421](https://github.com/realm/realm-js/issues/4421), since v10.12.0)
+* Fixed remaining uses of `globalThis` undefined issue, causing Realm to not load on iOS 11/12. ([#4350](https://github.com/realm/realm-js/issues/4350))
```
Comment (kneth) [inline CHANGELOG.md:184]: :shipit:  Do we know if it has always been an issue - or was the bug introduced in a specific version?

**Review by kneth — APPROVED** [review]

**Review by kraenhansen — COMMENTED** [review]
I'm wondering if we could simplify the solution, but simply documenting the use of the [@babel/preset-env](https://babeljs.io/docs/en/babel-preset-env#shippedproposals) Babel preset?

**Comment by tomduncalf** [comment]
> I'm wondering if we could simplify the solution, but simply documenting the use of the [@babel/preset-env](https://babeljs.io/docs/en/babel-preset-env#shippedproposals) Babel preset?

As in we say to users that they need to use `@babel/preset-env` in their project if they are on an old iOS version?

**Comment by kraenhansen** [comment]
> As in we say to users that they need to use @babel/preset-env in their project if they are on an old iOS version?

Yeah. Or perhaps https://www.npmjs.com/package/babel-plugin-transform-globalthis?

**Comment by kraenhansen** [comment]
Regardless, if we choose to discourages the explicit usage of `globalThis`, I think this PR should ensure `eslint` errors on that.

**Comment by tomduncalf** [comment]
I was following the pattern of what we've already done for `globalThis` in a couple of places, which is to fix our library so that it works with all supported OS versions. I _think_ I prefer that approach (we do the nasty stuff so users don't have to do any workarounds) but I am also happy if the majority would prefer to ask users to install a Babel plugin for this reasonably edge case.

@kneth @takameyer any thoughts?

Good suggestion re: the lint rule

**Comment by kraenhansen** [comment]
> we do the nasty stuff so users don't have to do any workarounds

I agree - and ideally we would apply this in a build step before publishing our package for consumption.
I'd expect this to be easier to implement in the future as I plan on adding rollup to the package as part of the binding generator project.

**Comment by tomduncalf** [comment]
>we would apply this in a build step before publishing our package for consumption.
Got it, this would be the ideal (we use the Babel plugin on our code before publish). In the interim until we have the infra in place to do that, are you OK with this solution?

**Comment by takameyer** [comment]
> any thoughts?
I think we should do the work for our users here.  If it's possible that we implement the babel plugin in our own codebase, I would prefer this.

**Review by takameyer — APPROVED** [review]
LGTM

**Comment by kneth** [comment]
> any thoughts?

I agree with @takameyer

**Comment by tomduncalf** [comment]
@tomduncalf Plan: merge this PR, rebase v11 against master, rebase v11 PR to remove cherrypick commit and merge

**Comment by tomduncalf** [comment]
If everyone is happy with this, I'll publish the new common version and rerun CI

**Comment by kraenhansen** [comment]
> I added the eslint rule in all public facing packages: b92a7f8

I think it should be enough to add it to the root package and have the other packages inherit?

**Comment by tomduncalf** [comment]
> I think it should be enough to add it to the root package and have the other packages inherit?

Ah I didn't realist that they inherited! Even better, I'll update

**Pull request merged**